File: Assets/Game/Scripts/GamePlay/GamePreviewManager.lua
---@gendoc

local SkyBoxUtils = Ballance2.Game.Utils.SkyBoxUtils
local KeyCode = UnityEngine.KeyCode
local GameSoundType = Ballance2.Services.GameSoundType

---关卡预览管理器，负责关卡预览模式时的一些控制行为。
---@class GamePreviewManager : GameLuaObjectHostClass
---@field GamePhysicsWorld PhysicsEnvironment
---@field GamePreviewCamera Camera
---@field GamePreviewMinimapCamera Camera
---@field GamePreviewFreeCamera FreeCamera
---@field GamePreviewCameraSkyBox Skybox
---@field GameDepthTestCubes GameObject[]
GamePreviewManager = ClassicObject:extend()

function GamePreviewManager:new()
  self._IsGamePlaying = true
  self._IsLoaded = false
  GamePlay.GamePreviewManager = self
end
function GamePreviewManager:Awake()
  local Mediator = Game.Mediator

  --注册全局事件
  Mediator:SubscribeSingleEvent(Game.CorePackage, "CoreGamePreviewManagerInitAndStart", 'GamePreviewManager', function (evtName, params)
    Game.SoundManager:PlayFastVoice('core.sounds:Misc_StartLevel.wav', GameSoundType.Normal) --播放开始音乐
    GameUI.GamePreviewUI.gameObject:SetActive(true)
    GameUI.GamePreviewUI:SetLevelInfo(params[1], params[2], params[3])
    GameUI.GamePreviewUI:SetFreeCamera(self.GamePreviewFreeCamera)
    GamePlay.CamManager.gameObject:SetActive(false)
    GamePlay.MusicManager:EnableBackgroundMusic()
    self.GamePreviewMinimapCamera.gameObject:SetActive(true)
    self.GamePreviewCamera.gameObject:SetActive(true)
    GamePlay.SectorManager:ActiveAllModulsForPreview() ---激活机关

    Game.UIManager:CloseAllPage()
    Game.UIManager:MaskBlackFadeOut(1)
    self._IsLoaded = true

    --ESC键
    self.escKeyId = Game.UIManager:ListenKey(KeyCode.Escape, function (key, down)
      if down then
        if self._IsGamePlaying then
          self:PauseLevel()
        else
          self:ResumeLevel()
        end
      end
    end)
    return false
  end)
end
function GamePreviewManager:OnDestroy()
  local Mediator = Game.Mediator

  if self.escKeyId then
    Game.UIManager:DeleteKeyListen(self.escKeyId)
    self.escKeyId = nil
  end

  --取消注册全局事件
  Mediator:UnRegisterSingleEvent("CoreGamePreviewManagerInitAndStart")
end

function GamePreviewManager:PauseLevel()
  Game.SoundManager:PlayFastVoice('core.sounds:Menu_click.wav', GameSoundType.UI)
  Game.UIManager:GoPage('PageGamePreviewPause') 
  self._IsGamePlaying = false
end
function GamePreviewManager:QuitLevel()
  if self._IsLoaded then
    self._IsLoaded = false
    Game.SoundManager:PlayFastVoice('core.sounds:Menu_load.wav', GameSoundType.Normal)
    Game.UIManager:MaskBlackFadeIn(1)
    LuaTimer.Add(1000, function ()
      GameUI.GamePreviewUI.gameObject:SetActive(false)
      GamePlay.MusicManager:DisableBackgroundMusic()
      GamePlay.CamManager.gameObject:SetActive(true)
      Game.LevelBuilder:UnLoadLevel(nil)
      self.GamePreviewMinimapCamera.gameObject:SetActive(false)
      self.GamePreviewCamera.gameObject:SetActive(false)
    end)
  end
end
function GamePreviewManager:ResumeLevel()
  self._IsGamePlaying = true
  Game.SoundManager:PlayFastVoice('core.sounds:Menu_click.wav', GameSoundType.UI)
  Game.UIManager:CloseAllPage()
end

---初始化灯光和天空盒
---@param skyBoxPre string A-K 或者空，为空则使用 customSkyMat 材质
---@param customSkyMat Material 自定义天空盒材质
---@param lightColor Color 灯光颜色
function GamePreviewManager:CreateSkyAndLight(skyBoxPre, customSkyMat, lightColor)
  self.GamePreviewCameraSkyBox.material = customSkyMat or SkyBoxUtils.MakeSkyBox(skyBoxPre) --Init sky
  Ballance2.Services.GameManager.GameLight.color = lightColor
end
--隐藏天空盒和关卡灯光
function GamePreviewManager:HideSkyAndLight()
  self.GamePreviewCameraSkyBox.material = nil
end

function GamePreviewManager:ToggleDethTethCubesVisible(visible)
  for _, value in ipairs(self.GameDepthTestCubes) do
    value:SetActive(visible)
  end
end
function GamePreviewManager:ToggleSkylayerVisible(visible)
  if Game.LevelBuilder._CurrentLevelSkyLayer then
    Game.LevelBuilder._CurrentLevelSkyLayer:SetActive(visible)
  end
end

function CreateClass:GamePreviewManager() return GamePreviewManager() end

File: Assets/Game/GamePlay/Scripts/Preview/GamePreviewManager.cs
using Ballance2.Base;
using Ballance2.Game.Utils;
using Ballance2.Menu;
using Ballance2.Package;
using Ballance2.Services;
using BallancePhysics.Wapper;
using UnityEngine;
using UnityEngine.InputSystem;

namespace Ballance2.Game.GamePlay
{
  /// <summary>
  /// 关卡预览管理器，负责关卡预览模式时的一些控制行为。
  /// </summary>
  public class GamePreviewManager : GameSingletonBehavior<GamePreviewManager>
  {
    private const string TAG = "GamePreviewManager";

    private bool _IsGamePlaying = false;
    private bool _IsLoaded = false;

    public Camera GamePreviewCamera;
    public Camera GamePreviewMinimapCamera;
    public FreeCamera GamePreviewFreeCamera;
    public Skybox GamePreviewCameraSkyBox;
    [HideInInspector]
    public GameObject[] GameDepthTestCubes;
    [HideInInspector]
    public MusicManager MusicManager;
    [HideInInspector]
    public CamManager CamManager;
    [HideInInspector]
    public SectorManager SectorManager;

    private GameMediator Mediator;
    private GameUIManager UIManager;
    private GameSoundManager SoundManager;
    private GamePlayPreviewUIControl GamePreviewUI;
    [SerializeField]
    private InputAction ActionPause;

    private void Awake() {
      Mediator = GameMediator.Instance;
      UIManager = GameUIManager.Instance;
      SoundManager = GameSoundManager.Instance;
      var SystemPackage = GamePackage.GetSystemPackage();

      //注册全局事件
      Mediator.SubscribeSingleEvent(SystemPackage, "CoreGamePreviewManagerInitAndStart", "GamePreviewManager", (evtName, param) => {
        GamePreviewUI = GamePlayPreviewUIControl.Instance;
        SoundManager.PlayFastVoice("core.sounds:Misc_StartLevel.wav", GameSoundType.Normal); //播放开始音乐
        GamePreviewUI.gameObject.SetActive(true);
        GamePreviewUI.SetLevelInfo(param[0] as string, param[1] as string, param[2] as string);
        GamePreviewUI.SetFreeCamera(GamePreviewFreeCamera);
        CamManager.gameObject.SetActive(false);
        CamManager._CamTarget.gameObject.SetActive(false);
        MusicManager.EnableBackgroundMusic();
        GamePreviewMinimapCamera.gameObject.SetActive(true);
        GamePreviewCamera.gameObject.SetActive(true);
        SectorManager.ActiveAllModulsForPreview(); //激活机关

        UIManager.CloseAllPage();
        UIManager.MaskBlackFadeOut(1);
        _IsLoaded = true;

        ActionPause.Enable();
        ActionPause.performed += (context) => {
          if (context.ReadValueAsButton()) {
            if (_IsGamePlaying)
              PauseLevel();
            else
              ResumeLevel();
          }
        };
        return false;
      });
    }
    protected override void OnDestroy() {
      ActionPause.Disable();
      //取消注册全局事件
      Mediator.UnRegisterSingleEvent("CoreGamePreviewManagerInitAndStart");
    }

    public void PauseLevel() {
      SoundManager.PlayFastVoice("core.sounds:Menu_click.wav", GameSoundType.UI);
      UIManager.GoPage("PageGamePreviewPause") ;
      _IsGamePlaying = false;
    }
    public void QuitLevel() {
      if (_IsLoaded) {
        _IsLoaded = false;
        SoundManager.PlayFastVoice("core.sounds:Menu_load.wav", GameSoundType.Normal);
        UIManager.MaskBlackFadeIn(1);
        GameTimer.Delay(1, () => {
          GamePreviewUI.gameObject.SetActive(false);
          MusicManager.DisableBackgroundMusic();
          CamManager.gameObject.SetActive(true);
          LevelBuilder.LevelBuilder.Instance.UnLoadLevel(null);
          GamePreviewMinimapCamera.gameObject.SetActive(false);
          GamePreviewCamera.gameObject.SetActive(false);
        });
      }
    }
    public void ResumeLevel() {
      _IsGamePlaying = true;
      SoundManager.PlayFastVoice("core.sounds:Menu_click.wav", GameSoundType.UI);
      UIManager.CloseAllPage();
    }

    /// <summary>
    /// 初始化灯光和天空盒
    /// </summary>
    /// <param name="skyBoxPre">A-K 或者空，为空则使用 customSkyMat 材质</param>
    /// <param name="customSkyMat">自定义天空盒材质</param>
    /// <param name="lightColor">灯光颜色</param>
    public void CreateSkyAndLight(string skyBoxPre, Material customSkyMat, Color lightColor) {
      GamePreviewCameraSkyBox.material = customSkyMat == null ? SkyBoxUtils.MakeSkyBox(skyBoxPre) : customSkyMat;
      Ballance2.Services.GameManager.GameLight.color = lightColor;
    }
    /// <summary>
    /// 隐藏天空盒和关卡灯光
    /// </summary>
    public void HideSkyAndLight() {
      GamePreviewCameraSkyBox.material = null;
    }

    public void ToggleDethTethCubesVisible(bool visible) {
      foreach (var value in GameDepthTestCubes)
        value.SetActive(visible);
    }
    public void ToggleSkylayerVisible(bool visible) {
      if (LevelBuilder.LevelBuilder.Instance.CurrentLevelSkyLayer)
        LevelBuilder.LevelBuilder.Instance.CurrentLevelSkyLayer.SetActive(visible);
    }
  }
}